### ISSUE TITLE
Unable to )clear after error deserialising an array

### ISSUE THREAD

**Issue by LdBeth** [Dyalog/link#344]
Even using `Link.Break` won't make `)clear` work again.

**To Reproduce**

```
$ pwd
/Users/ldbeth/Image
$ ls
dotmask.apla
$ cat dotmask.apla
[
  35 48 40 32 28 15 23 31
  43 59 56 52 20  4  7 11
  51 62 60 44 12  1  3 19
  38 46 54 36 25 17  9 27
  29 14 22 30 34 49 41 33
  21  5  6 10 42 48 57 53
  13  0  2 18 50 63 61 45
  24 16  8 26 39 47 55 37
]
```
**Session transcript**
```
      ]link.create foo /Users/ldbeth/Image
Linked: #.foo ←→ /Users/ldbeth/Image
      )clear
Cannot perform operation when # is referenced by session namespace.
⎕SE.Dyalog...Array...H
      ]link.status
 Namespace  Directory            Files 
 #.foo      /Users/ldbeth/Image      1 
      ]link.break #.foo
Unlinked: #.foo
      ]link.status
No active links
      )clear
Cannot perform operation when # is referenced by session namespace.
⎕SE.Dyalog...Array...H
```

**Desktop (please complete the following information):**
 - OS: macOS
 - APL version: Unicode Version 18.0.38756
 - `⎕SE.Link.Version`: 3.0.0, 3a86ded044f3c895d5e6748607d4c9503e71ae1c

**Comment by mkromberg** [comment]
I think I can reproduce this, can you confirm whether 1) you have a trailing blank line in the file, and 2) whether the problem goes away if you remove that line. If so, there are at least two bugs here, one in the array deserialiser and one in the error handling.

**Comment by LdBeth** [comment]
The file created by LINK has only a `\n` character after `]`. However, remove that `\n` does not prevent the problem.

**Comment by mkromberg** [comment]
Can you attach a copy of the file to this issue?

**Comment by LdBeth** [comment]
[dotmask.zip](https://github.com/Dyalog/link/files/7158763/dotmask.zip)

**Comment by LdBeth** [comment]
I find something interesting:

```
      ]link.create Foo /Users/ldbeth/Image
Linked: #.Foo ←→ /Users/ldbeth/Image
      )clear
Cannot perform operation when # is referenced by session namespace.
⎕SE.Dyalog...Array...H
      )cs ⎕SE.Dyalog.Array
⎕SE.Dyalog.Array
      )fns
Deserialise     DeserialiseQA   H       L       QA      RoundtripQA     
Serialise       SerialiseQA     ∆NS     ∆NSinverse      
      )ed H
      H
 #.Foo       ∇{⍺⍺⍎⍵⍵,'⍵'}
     ∇  {⎕ML←1⋄↑⍵}1/¨ 
      )erase H
      )clear
clear ws
```
I believe it is related to this line: https://github.com/Dyalog/link/blob/3a86ded044f3c895d5e6748607d4c9503e71ae1c/StartupSession/Dyalog/Array.apln#L229

I don't know very well on how LINK and session namespace works, but it seems the `H` has been "leaked" into the session namespace. Erase `H` can resolve the problem temporarily until next time the directory is linked. And if a source directory without array is linked, the `H` does not appear.

**Comment by mkromberg** [comment]
OK, I'm having trouble reproducing that. Are you able to show how to reproduce it from a newly started session? I don't see how H can become a function, as your output seems to be suggesting. I *can* see from the code that setting DEBUG←1 causes deserialisation to follow a different path, but I cannot get it to leave an H behind. It may be more efficient to contact me via e-mail (mkrom@...) while we debug this.

**Comment by LdBeth** [comment]
Calling `Deserialize` directly from a fresh session can also results in producing `H` here.

```
Wed Sep 15 12:33:14 2021
      ⎕SE.Dyalog.Array.DEBUG←1 ⍝ This is the default value set in the master branch
      ⎕SE.Dyalog.Array.Deserialise'[' '1 2 3 4' '5 6 7 8' ']'
1 2 3 4
5 6 7 8
      ⎕SE.Dyalog.Array.H
 ⎕SE.Dyalog.Array .  #       ∇{⍺⍺⍎⍵⍵,'⍵'}
     ∇  {⎕ML←1⋄↑⍵}1/¨  
      )ns New
#.New
      )cs New
#.New
      ⎕SE.Dyalog.Array.Deserialise'[' '1 2 3 4' '5 6 7 8' ']'
1 2 3 4
5 6 7 8
      ⎕SE.Dyalog.Array.H ⍝ The corresponding namespace has changed.
 ⎕SE.Dyalog.Array .  #.New       ∇{⍺⍺⍎⍵⍵,'⍵'}
     ∇  {⎕ML←1⋄↑⍵}1/¨  
      )clear ⍝ stuck
Cannot perform operation when # is referenced by session namespace.
⎕SE.Dyalog...Array...H
```

**Comment by LdBeth** [comment]
I think I find where exactly it goes wrong, it is possibly due to a bug in dfns error-guard, the direct cause is this line

https://github.com/Dyalog/link/blob/3a86ded044f3c895d5e6748607d4c9503e71ae1c/StartupSession/Dyalog/Array.apln#L228

The "error-guard inside guard" in this line somehow prevented the local variable cleanup happen, and the function returned with local variables leaked in to the namespace.

Curiously, make this guard into two lines would prevent the leaking, i.e.

```
⍺{0::0 ⋄ r←⍺⍎⍵ ⋄ ~(⊃⎕NC'r')∊3 4}h:
⍺⍎⍵
```

Should I fire a bug to the upstream?

**Comment by LdBeth** [comment]
To reproduce the above bug,

1. start with a new session
2. edit `⎕SE.Dyalog.Array.Deserialise` and set a break point at that line. 
<img width="529" alt="截圖 2021-09-15 下午5 25 03" src="https://user-images.githubusercontent.com/22933696/133408296-4ee1c165-1020-47d3-9e4c-3238f76a8ec6.png">

3. execute `⎕SE.Dyalog.Array.Deserialise'[' '1 2 3 4' ']'`, let the trace window pops up.
4. repeating step into the function, and finally would find the function returned at that line but the local variable cleanup didn't happen.

**Comment by lstefano71** [comment]
` 
foo←{1:1}
`
and 
```
 goo←{1:
     1} 
```
are two very different functions.
Try:
```
     1≡foo ''
1
```
vs.
```
       1≡goo ''
VALUE ERROR: No result was provided when the context expected one
      1≡goo''                                                    
        ∧                                                        
```

**Comment by LdBeth** [comment]
@lstefano71 right I made a mistake there.

This is the minimal example I found can reproduce the variable leak bug.

```
 ExecuteEach←{
     BBA←'BBA' ⍝ Test
     ABC←1 2 3
     GGC←{⍵+1}
     ⍝ Any variables before this point will leak
     {0::0 ⋄ r←⍎⍵ ⋄ 1}⍵:1 ⍝ execute ⍵
     ⍝ Enclose is needed to prevent tail recursion.
     ⊂∇'{×⍵}1' ⍝ a dfns is needed to trigger error
 }

    ExecuteEach ']'
1
      )vars
ABC     BBA     
      )fns
ExecuteEach     GGC
```

**Comment by mkromberg** [comment]
I am unable to reproduce this with the my copy of 18.0 (revision 40684). Is it possible for you to upgrade your interpreter and try again? 18.0.4684 is available via the Download button on our web page. I know that bugs were found and fixed in this area.

**Comment by LdBeth** [comment]
Yes it has been fixed in the `18.0.40639` I just downloaded.
